src/bsoncxx/lib/bsoncxx/v1/document/view.cpp

```diff
@@ -30,6 +43,167 @@ static_assert(
+        return this->cend();
+    }
+
+    bson_iter_t iter;
+
+    if (!bson_iter_init_from_data(&iter, _data, this->size())) {
+        throw v1::exception{code::invalid_data};
+    }
+
+    if (!bson_iter_next(&iter)) {
+        return this->cend();
+    }
+
+    return const_iterator::internal::make_const_iterator(
+        _data, this->size(), bson_iter_offset(&iter), bson_iter_key_len(&iter));
+}
+
+view::const_iterator view::find(v1::stdx::string_view key) const {
+    if (!this->operator bool()) {
+        return this->cend();
+    }
+
+    if (key.size() >= std::size_t{INT_MAX}) {
+        return this->cend();
+    }
+
+    // Support null as equivalent to empty.
+    if (!key.data()) {
+        key = "";
+    }
+
+    bson_t bson;
+
+    if (!bson_init_static(&bson, _data, this->size())) {
+        throw v1::exception{code::invalid_data};
+    }
+
+    bson_iter_t iter;
+
+    if (!bson_iter_init_find_w_len(&iter, &bson, key.data(), static_cast<int>(key.size()))) {
+        return this->end();
+    }
+
+    return const_iterator::internal::make_const_iterator(
+        _data, this->size(), bson_iter_offset(&iter), bson_iter_key_len(&iter));
+}
+
+std::error_category const& view::error_category() {
+    class type final : public std::error_category {
+        char const* name() const noexcept override {
+            return "bsoncxx::v1::document::view";
+        }
+
+        std::string message(int v) const noexcept override {
+            switch (static_cast<code>(v)) {
+                case code::zero:
+                    return "zero";
+                case code::invalid_length:
+                    return "length is invalid";
+                case code::invalid_data:
+                    return "data is invalid";
+                default:
+                    return "unknown: " + std::to_string(v);
+            }
+        }
+
+        bool equivalent(int v, std::error_condition const& ec) const noexcept override {
+            if (ec.category() == v1::source_error_category()) {
+                using condition = v1::source_errc;
+
+                auto const source = static_cast<condition>(ec.value());
+
+                switch (static_cast<code>(v)) {
+                    case code::invalid_length:
+                    case code::invalid_data:
+                        return source == condition::bsoncxx;
+
+                    case code::zero:
+                    default:
+                        return false;
+                }
+            }
+
+            if (ec.category() == v1::type_error_category()) {
+                using condition = v1::type_errc;
+
+                auto const type = static_cast<condition>(ec.value());
+
+                switch (static_cast<code>(v)) {
+                    case code::invalid_length:
+                        return type == condition::invalid_argument;
+
+                    case code::invalid_data:
+                        return type == condition::runtime_error;
```

Comment:
As a reminder, `type_errc::invalid_argument` is meant to only apply to immediate arguments given to the invoked function from which the exception may be thrown. Here, because `invalid_data` is only thrown when `bson_*` functions report failure (not by checking the value of an immediate function parameter), it is categorized as a `type_errc::runtime_error`. For this purpose, data members of `this` are not considered to be "arguments" to the given member function. This rationale applies to all other v1 error codes.